packages/outbox-core/lib/outbox.ts

```diff
@@ -0,0 +1,127 @@
+import { AbstractPeriodicJob, type JobExecutionContext } from '@lokalise/background-jobs-common'
+import type { PeriodicJobDependencies } from '@lokalise/background-jobs-common'
+import type {
+  CommonEventDefinition,
+  CommonEventDefinitionPublisherSchemaType,
+  ConsumerMessageMetadataType,
+  DomainEventEmitter,
+} from '@message-queue-toolkit/core'
+import { PromisePool } from '@supercharge/promise-pool'
+import { uuidv7 } from 'uuidv7'
+import type { OutboxAccumulator } from './accumulators'
+import type { OutboxStorage } from './storage'
+
+/**
+ * Main logic for handling outbox entries.
+ *
+ * If entry is rejected, it is NOT going to be handled during the same execution. Next execution will pick it up.
+ */
+export class OutboxProcessor<SupportedEvents extends CommonEventDefinition[]> {
+  constructor(
+    private readonly outboxStorage: OutboxStorage<SupportedEvents>,
+    private readonly outboxAccumulator: OutboxAccumulator<SupportedEvents>,
+    private readonly eventEmitter: DomainEventEmitter<SupportedEvents>,
+    private readonly maxRetryCount: number,
+    private readonly emitBatchSize: number,
+  ) {}
+
+  public async processOutboxEntries(context: JobExecutionContext) {
+    const entries = await this.outboxStorage.getEntries(this.maxRetryCount)
+
+    await PromisePool.for(entries)
+      .withConcurrency(this.emitBatchSize)
+      .process(async (entry) => {
+        try {
+          await this.eventEmitter.emit(entry.event, entry.data, entry.precedingMessageMetadata)
+          await this.outboxAccumulator.add(entry)
+        } catch (e) {
```

Comment:
🟢 It's a matter of taste, but you can rely on PromisePool for giving you a list of errors:

```ts
const { results, errors } = await PromisePool
  .for([1, 2, 3])
  .withConcurrency(5)
  .useCorrespondingResults()
  .process(async (number) => {
    // …
  })
```

that way try-catch be eliminated. nothing wrong with it, also just a matter of preference

Reply:
Impemented promise pool.